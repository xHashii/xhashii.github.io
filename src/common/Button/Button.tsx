import React, { FC, ReactChildren, ReactElement, DetailedHTMLProps, HTMLAttributes } from "react";
import { Classes } from "jss";
import classNames from "classnames";

export interface IButtonProps extends DetailedHTMLProps<HTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  classes: Classes<"button" | "flat" | "secondary" | "buttonIcon">;
  secondary?: boolean;
  flat?: boolean;
  children?: ReactChildren | string;
  icon?: ReactElement;
  success?: boolean;
  danger?: boolean;
  disabled?: boolean;
}

const Button: FC<IButtonProps> = ({
  classes = {},
  secondary = false,
  flat = false,
  children = null,
  icon = null,
  success = false,
  danger = false,
  disabled = false,
  className = "",
  ...props
}) => {
  let cn = classes.button;

  if (secondary) cn = classes.secondary;

  if (flat) cn = classes.flat;

  return (
    <button disabled={disabled} {...props} className={classNames(cn, className)}>
      { children }
      { icon && React.cloneElement(icon, { className: classes.buttonIcon }) }
    </button>
  );
};

export default Button;
