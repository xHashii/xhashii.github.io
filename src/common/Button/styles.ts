import theme, { Theme } from "../../utils/theme";
import { IButtonProps } from "./Button";

export const defaultButtonStyle = {
  display: "inline-block",
  textDecoration: "none",
  textAlign: "center" as "center",
  position: "relative" as "relative",
  border: 0,
  WebkitBoxShadow: "0 1px 5px 0 rgba(0,0,0,0.06)",
  boxShadow: "0 1px 5px 0 rgba(0,0,0,0.06)",
  fontSize: 16,
  outline: "none",
  fontWeight: "bold" as "bold",
  minWidth: 88,
  margin: 8,
  borderRadius: 4,
  cursor: "pointer",
  fontFamily: theme.fontFamily,
  padding: "12px 32px"
};

export default (theme: Theme) => ({
  [theme.media.mobile as any]: {
    button: {
      width: "100%"
    },
    secondary: {
      width: "100%"
    },
    link: {
      width: "calc(100% - 66px)"
    }
  },
  button: {
    ...defaultButtonStyle,
    background: ({ danger, success, disabled }: IButtonProps): string => {
      if (disabled) return theme.colorGrey;

      if (danger) return theme.colorRed;

      if (success) return theme.colorGreen;

      return theme.colorActive;
    },
    border: 0,
    color: "#fff",
    textShadow: "0 1px 0 rgba(0,0,0,0.03)"
  },
  secondary: {
    ...defaultButtonStyle,
    backgroundColor: "transparent",
    border: `1px solid ${theme.colorActive}`,
    color: theme.colorActive
  },
  flat: {
    ...defaultButtonStyle,
    color: ({ disabled }: IButtonProps) => disabled ? theme.colorGrey : theme.colorActive,
    border: 0,
    WebkitBoxShadow: "0",
    boxShadow: "0",
    background: "transparent"
  },
  buttonIcon: {
    fontSize: 20,
    marginLeft: 8,
    marginBottom: -4
  }
});
