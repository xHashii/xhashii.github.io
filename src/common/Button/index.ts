import withSyles from "react-jss";

import Button from "./Button";
import styles from "./styles";

export default withSyles(styles)(Button);
