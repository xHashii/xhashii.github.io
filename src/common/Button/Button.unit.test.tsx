import React from "react";
import { shallow } from "enzyme";

import Button from "./Button";

const defaultProps = {
  classes: {
    button: "",
    secondary: "",
    flat: "",
    buttonIcon: ""
  }
};

describe("The Button instance", () => {
  describe("when called with default properties", () => {
    it("should render without throwing errors", () => {
      const wrapper = shallow(<Button {...defaultProps} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when the onClick prop is passed to the button", () => {
    describe("it should be called when the button is clicked", () => {
      const newProps = {
        ...defaultProps,
        onClick: jest.fn(),
        success: true
      };
      const wrapper = shallow(<Button {...newProps} />);
      const button = wrapper.find("button");

      button.simulate("click");

      expect(newProps.onClick).toHaveBeenCalledTimes(1);
    });
  });

  describe("when style overrides are passed", () => {
    it("should add a style attribute to the element with the overrides", () => {
      const newProps = {
        ...defaultProps,
        danger: true,
        style: {
          fontSize: 0
        }
      };
      const wrapper = shallow(<Button {...newProps} />);
      const button = wrapper.find("button");

      expect(button.props().style).toEqual(newProps.style);
    });
  });

  describe("when the secondary prop is passed", () => {
    it("should render a secondary button", () => {
      const newProps = {
        ...defaultProps,
        secondary: true
      };
      const wrapper = shallow(<Button {...newProps} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when the icon prop is passed", () => {
    it("should render the icon inside the button", () => {
      const newProps = {
        ...defaultProps,
        icon: <div data-test-id="icon" />
      };
      const wrapper = shallow(<Button {...newProps} />);
      const icon = wrapper.find({ "data-test-id": "icon" });

      expect(icon.exists()).toBe(true);
    });
  });
});
