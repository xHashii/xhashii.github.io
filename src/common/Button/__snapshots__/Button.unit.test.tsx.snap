// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Button instance when called with default properties should render without throwing errors 1`] = `ShallowWrapper {}`;

exports[`The Button instance when the secondary prop is passed should render a secondary button 1`] = `ShallowWrapper {}`;
